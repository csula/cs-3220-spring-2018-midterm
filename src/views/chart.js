export class GameChart extends HTMLElement {
    constructor () {
        super();
        this._values = [];
        this._color = '#ff0000';
        this._displayStyle = 'horizontal';
    }

    static get baseUnit () {
        return 16;
    }
    // this allows GameChart to listen for dom attribute changes and call
    // attributeChangedCallback when the attribute changes in DOM
    static get observedAttributes () {
        return ['data-values', 'data-color', 'data-display-style'];
    }

    _upgradeProperty(prop) {
        if (this.hasOwnProperty(prop)) {
            let value = this[prop];
            delete this[prop];
            this[prop] = value;
        }
    }

    /**
     * define custom getters and setters to handle update on value changes
     * accoridngly through JavaScript call
     */
    get values () {
        return this._values;
    }
    set values (val) {
        this._values = val;
        this.handleChange();
    }
    get color () {
        return this._color;
    }
    set color (val) {
        this._color = val;
        this.handleChange();
    }
    get displayStyle () {
        return this._displayStyle;
    }
    set displayStyle (val) {
        this._displayStyle = val;
        this.handleChange();
    }

    connectedCallback () {
        // upgrade property so that the es6 setter can be called properly
        ['colors', 'values', 'displayStyle'].forEach(p => {
            this._upgradeProperty(p);
        });
        // TODO: call render method :)
    }

    handleChange() {
        // TODO: handle update and re-render
    }

    attributeChangedCallback () {
        // TODO: get attributes from dataset and re-render accordingly
    }

    render () {
        // TODO: render chart DOM
        // note: the bar should display in *reverse* order of the values
    }
}

