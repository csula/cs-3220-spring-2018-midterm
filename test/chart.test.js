/*eslint-env node, jest */

import { GameChart } from '../src/views/chart';
require('@skatejs/ssr/register');
const render = require('@skatejs/ssr');

describe('Chart Component', () => {
    beforeEach(() => {
        window.customElements.define('game-chart', GameChart);
    });

    it('should display empty chart', done => {
        render(new GameChart()).then(html => {
            expect(html).toMatchSnapshot();
            done();
        });
    });

    it('should display chart with bar', done => {
        const chart = new GameChart();
        chart.values = [1, 2, 3];
        render(chart).then(html => {
            expect(html).toMatchSnapshot();
            done();
        });
    });

    it('should be able to update color', done => {
        const chart = new GameChart();
        chart.values = [1, 2, 3];
        render(chart).then(html => {
            expect(html).toMatchSnapshot();
        })
        .then(() => {
            chart.color = '#00ff00';
            return render(chart);
        }).then(html => {
            expect(html).toMatchSnapshot();
            done();
        });
    });
});
