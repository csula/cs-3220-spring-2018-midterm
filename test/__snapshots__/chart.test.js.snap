// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component should be able to update color 1`] = `

<game-chart>
  <div class="chart-container">
    <div class="bar"
         style="width: 48px; background-color: #ff0000;"
    >
      3
    </div>
    <div class="bar"
         style="width: 32px; background-color: #ff0000;"
    >
      2
    </div>
    <div class="bar"
         style="width: 16px; background-color: #ff0000;"
    >
      1
    </div>
  </div>
</game-chart>

`;

exports[`Chart Component should be able to update color 2`] = `

<game-chart>
  <div class="chart-container">
    <div class="bar"
         style="width: 48px; background-color: #00ff00;"
    >
      3
    </div>
    <div class="bar"
         style="width: 32px; background-color: #00ff00;"
    >
      2
    </div>
    <div class="bar"
         style="width: 16px; background-color: #00ff00;"
    >
      1
    </div>
  </div>
</game-chart>

`;

exports[`Chart Component should display chart with bar 1`] = `

<game-chart>
  <div class="chart-container">
    <div class="bar"
         style="width: 48px; background-color: #ff0000;"
    >
      3
    </div>
    <div class="bar"
         style="width: 32px; background-color: #ff0000;"
    >
      2
    </div>
    <div class="bar"
         style="width: 16px; background-color: #ff0000;"
    >
      1
    </div>
  </div>
</game-chart>

`;

exports[`Chart Component should display empty chart 1`] = `

<game-chart>
  <div class="chart-container">
  </div>
</game-chart>

`;
